kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala

```diff
@@ -673,6 +673,76 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
     }
   }
 
+  test("insert into table - idempotent writes") {
+    withTempDirAndEngine { (tblPath, engine) =>
+      val data = Seq(Map("part1" -> ofInt(1), "part2" -> ofInt(2)) -> dataPartitionBatches1)
+      var expData = Seq.empty[TestRow] // as the data in inserted, update this.
+
+      def addDataWithTxnId(newTbl: Boolean, appId: String, txnVer: Long, expTblVer: Long): Unit = {
+        var txnBuilder = createWriteTxnBuilder(Table.forPath(engine, tblPath))
+
+        if (appId != null) txnBuilder = txnBuilder.withTransactionId(engine, appId, txnVer)
+
+        if (newTbl) {
+          txnBuilder = txnBuilder.withSchema(engine, testPartitionSchema)
+            .withPartitionColumns(engine, testPartitionColumns.asJava)
+        }
+        val txn = txnBuilder.build(engine)
+
+        val combinedActions = inMemoryIterable(
+          data.map { case (partValues, partData) =>
+            stageData(txn.getTransactionState(engine), partValues, partData)
+          }.reduceLeft(_ combine _))
+
+        val commitResult = txn.commit(engine, combinedActions)
+
+        expData = expData ++ data.flatMap(_._2).flatMap(_.toTestRows)
+
+        verifyCommitResult(commitResult, expVersion = expTblVer, expIsReadyForCheckpoint = false)
+        val expPartCols = if (newTbl) testPartitionColumns else null
+        verifyCommitInfo(tblPath, version = expTblVer, expPartCols, operation = WRITE)
+        verifyWrittenContent(tblPath, testPartitionSchema, expData)
+      }
+
+      def expFailure(appId: String, txnVer: Long, latestTxnVer: Long)(fn: => Any): Unit = {
+        val ex = intercept[ConcurrentTransactionException] {
+          fn
+        }
+        assert(ex.getMessage.contains(s"This error occurs when multiple updates are using the " +
```

Comment:
cc @allisonport-db - what is our final decision with error codes? Checking errror codes would be a lot nicer/cleaner in tests

Reply:
I think we decided to punt on the error codes and add them later if the need arises.